lib/rules/lint-no-whitespace-within-word.js

```diff
@@ -0,0 +1,66 @@
+// lint-no-whitespace-within-word.js
+'use strict';
+const Rule = require('./base');
+const createErrorMessage = require('../helpers/create-error-message');
+const ERROR_MESSAGE = 'Excess whitespace in layout detected.';
+
+// Entity format: nbsp -> &nbsp; -> '\\&nbsp\\;'
+// Number format: 160 -> &#160; -> '\\&\\#160\\;'
+const blackList = [
+  '\\&\\#32\\;', ' ',
+  '\\&\\#160\\;', '\\&nbsp\\;', '\\&NonBreakingSpace\\;',
+  '\\&\\#8194\\;', '\\&ensp\\;',
+  '\\&\\#8195\\;', '\\&emsp\\;',
+  '\\&\\#8196\\;', '\\&emsp13\\;',
+  '\\&\\#8197\\;', '\\&emsp14\\;',
+  '\\&\\#8199\\;', '\\&numsp\\;',
+  '\\&\\#8200\\;', '\\&puncsp\\;',
+  '\\&\\#8201\\;', '\\&thinsp\\;', '\\&ThinSpace\\;',
+  '\\&\\#8202\\;', '\\&hairsp\\;', '\\&VeryThinSpace\\;', '\\&ThickSpace\\;',
+  '\\&\\#8203\\;', '\\&ZeroWidthSpace\\;', '\\&NegativeVeryThinSpace\\;', '\\&NegativeThinSpace\\;', '\\&NegativeMediumSpace\\;', '\\&NegativeThickSpace\\;',
+  '\\&\\#8204\\;', '\\&zwnj\\;',
+  '\\&\\#8205\\;', '\\&zwj\\;',
+  '\\&\\#8206\\;', '\\&lrm\\;',
+  '\\&\\#8207\\;', '\\&rlm\\;',
+  '\\&\\#8287\\;', '\\&MediumSpace\\;', '\\&ThickSpace\\;',
+  '\\&\\#8288\\;', '\\&NoBreak\\;',
+  '\\&\\#8289\\;', '\\&ApplyFunction\\;', '\\&af\\;',
+  '\\&\\#8290\\;', '\\&InvisibleTimes\\;', '\\&it\\;',
+  '\\&\\#8291\\;', '\\&InvisibleComma\\;', '\\&ic\\;'
+];
+
+// Join blackList entries into String as first capture group 
+// [ blackList[0], blackList[1], ..., blackList[n] ] ->
+// ((blackList[0])|(blackList[1])|...|(blackList[n]))
+const blackListString = '\(\(' + blackList.join('\)|\(') + '/)\)';
+
+// Second capture group: single non-whitepace character (\S)
+const whiteListString = '\(\\S\)';
+
+// Goal is to catch alternating blackList/whiteList 
+// characters, for example, in 'W e l c o m e'
+// So the final pattern boils down to 
+// (anyBlackListItem)(anyWhiteListItem)(anyBlackListItem)(anyWhiteListItem)(anyBlackListItem)
+const blackListRegExp = new RegExp(blackListString+whiteListString+blackListString+whiteListString+blackListString, 'g');
+
+
+module.exports = class noWhitespaceWithinWord extends Rule {
+  visitor() {
+    return {
+      TextNode(node) {
+        let source = this.sourceForNode(node);
+        let matches = source.match(blackListRegExp);
```

Comment:
I find this mega regexp hard to reason about. It might be easier to reason about this as a loop, where you keep track of the last seen item (whitespace vs non-whitespace) and then increment a counter if the next char is different. 

Something like

```js
let alternationCount = 0;
let source = this.sourceForNode(node);

for (let i = 0; i < source.length; i++) {
  let currentChar = source[i];
  let previousChar = i > 0 ? source[i - 1] : undefined;
  
  if (
    (isWhitespace(currentChar) && !isWhitespace(previousChar)) || 
    (!isWhitespace(currentChar) && isWhitespace(previousChar))
  ) {
    alternationCount++;
  }
}

if (alternationCount >= 3) {
 this._log({ ... });
}
```

Reply:
This seems like it will be slower but I can try it.

Reply:
> This seems like it will be slower but I can try it.

Hmm, based on what?

Reply:
I think this approach doesn't work, I tried it a few different ways - I had to timebox it because the submitted approach does function as intended. Let's add this to our review session and dig in a bit more.

Reply:
We resolved this together, updated the code below.